### PR TITLE
[extractor/odnoklassniki] Fix formats extraction

### LINKED ISSUE #2959: [ok.ru] request download video from webm format, save size/quality
### Checklist

- [X] I'm reporting a site feature request
- [X] I've verified that I'm running yt-dlp version **2022.02.04**. ([update instructions](https://github.com/yt-dlp/yt-dlp#update))
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [ ] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

_No response_

### Example URLs

https://ok.ru/video/3812509420088

### Description

Hello, default get format from ok.ru
```
C:\Users\Administrator>youtube-dl -F "https://ok.ru/video/3812509420088" --user-agent "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0"
[Odnoklassniki] 3812509420088: Downloading desktop webpage
[Odnoklassniki] 3812509420088: Downloading m3u8 information
[info] Available formats for 3812509420088:
ID       EXT RESOLUTION FPS │    FILESIZE   TBR PROTO  │ VCODEC        VBR ACODEC    ABR     ASR MORE INFO
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
mobile   mp4 unknown        │                   https  │ unknown           unknown
hls-208  mp4 256x106     25 │ ~ 231.93MiB  208k m3u8_n │ unknown      208k unknown    0k
mpd-0    mp4 256x106     25 │ ~ 136.89MiB  122k https  │ avc1.64000B  122k mp4a.40.2  0k 44100Hz DASH video, mp4_dash
lowest   mp4 unknown        │                   https  │ unknown           unknown
hls-434  mp4 426x178     25 │ ~ 484.87MiB  434k m3u8_n │ unknown      434k unknown    0k
mpd-1    mp4 426x178     25 │ ~ 400.45MiB  359k https  │ avc1.64000D  359k mp4a.40.2  0k 44100Hz DASH video, mp4_dash
low      mp4 unknown        │                   https  │ unknown           unknown
hls-903  mp4 640x266     25 │ ~1007.17MiB  903k m3u8_n │ unknown      903k unknown    0k
mpd-2    mp4 640x266     25 │ ~ 853.57MiB  765k https  │ avc1.4D4015  765k mp4a.40.2  0k 44100Hz DASH video, mp4_dash
sd       mp4 unknown        │                   https  │ unknown           unknown
hls-1496 mp4 852x356     25 │ ~   1.63GiB 1496k m3u8_n │ unknown     1496k unknown    0k
mpd-3    mp4 852x356     25 │ ~   1.48GiB 1361k https  │ avc1.4D401E 1361k mp4a.40.2  0k 44100Hz DASH video, mp4_dash
hd       mp4 unknown        │                   https  │ unknown           unknown
hls-4148 mp4 1280x534    25 │ ~   4.52GiB 4148k m3u8_n │ unknown     4148k unknown    0k
mpd-4    mp4 1280x534    25 │ ~   4.33GiB 3976k https  │ avc1.4D401F 3976k mp4a.40.2  0k 44100Hz DASH video, mp4_dash
```
And when you view-source, you will get [metadataWebmUrl]() download video from it can save size and quality for some reason.
Screenshot_6

Hope you can add this to extract_url function. Thank you.

Regards,

### Verbose log

```shell
C:\Users\Administrator>youtube-dl -FvU "https://vd196.mycdn.me/?expires=1646557772885&srcIp=116.xxx.xxx.161&pr=10&srcAg=GECKO&ms=45.136.22.4&type=4&sig=hLJz8ZaZSIk&ct=6&urls=185.226.53.6&clientType=0&id=2254678985272" --user-agent "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0"
[debug] Command-line config: ['-FvU', 'https://vd196.mycdn.me/?expires=1646557772885&srcIp=116.xxx.xxx.161&pr=10&srcAg=GECKO&ms=45.136.22.4&type=4&sig=hLJz8ZaZSIk&ct=6&urls=185.226.53.6&clientType=0&id=2254678985272', '--user-agent', 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0']
[debug] Encodings: locale cp1252, fs utf-8, out utf-8, err utf-8, pref cp1252
[debug] yt-dlp version 2022.02.04 [c1653e9] (win_exe)
[debug] Python version 3.8.10 (CPython 64bit) - Windows-10-10.0.19042-SP0
[debug] exe versions: ffmpeg 5.0-full_build-www.gyan.dev (setts), ffprobe 4.4.1-full_build-www.gyan.dev
[debug] Optional libraries: Cryptodome, mutagen, sqlite, websockets
[debug] Proxy map: {}
Latest version: 2022.02.04, Current version: 2022.02.04
yt-dlp is up to date (2022.02.04)
[debug] [generic] Extracting URL: https://vd196.mycdn.me/?expires=1646557772885&srcIp=116.xxx.xxx.161&pr=10&srcAg=GECKO&ms=45.136.22.4&type=4&sig=hLJz8ZaZSIk&ct=6&urls=185.226.53.6&clientType=0&id=2254678985272
[generic] ?expires=1646557772885&srcIp=116.xxx.xxx.161&pr=10&srcAg=GECKO&ms=45.136.22.4&type=4&sig=hLJz8ZaZSIk&ct=6&urls=185.226.53: Requesting header
WARNING: [generic] Falling back on generic information extractor.
[generic] ?expires=1646557772885&srcIp=116.xxx.xxx.161&pr=10&srcAg=GECKO&ms=45.136.22.4&type=4&sig=hLJz8ZaZSIk&ct=6&urls=185.226.53: Downloading webpage
[generic] ?expires=1646557772885&srcIp=116.xxx.xxx.161&pr=10&srcAg=GECKO&ms=45.136.22.4&type=4&sig=hLJz8ZaZSIk&ct=6&urls=185.226.53: Extracting information
[debug] Identified a DASH manifest
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
WARNING: Extractor failed to obtain "title". Creating a generic title instead
[info] Available formats for ?expires=1646557772885&srcIp=116.xxx.xxx.161&pr=10&srcAg=GECKO&ms=45.136.22.4&type=4&sig=hLJz8ZaZSIk&ct=6&urls=185.226.53:
ID EXT  RESOLUTION FPS │   TBR PROTO │ VCODEC       VBR ACODEC      ABR     ASR MORE INFO
──────────────────────────────────────────────────────────────────────────────────────────────────────
5  webm audio only     │   32k https │ audio only       opus        32k 48000Hz DASH audio, webm_dash
6  webm audio only     │  128k https │ audio only       opus       128k 48000Hz DASH audio, webm_dash
7  webm audio only     │  192k https │ audio only       opus       192k 48000Hz DASH audio, webm_dash
0  webm 256x106     25 │  158k https │ vp9         158k video only              DASH video, webm_dash
2  webm 426x178     25 │  164k https │ vp9         164k video only              DASH video, webm_dash
1  webm 640x266     25 │  279k https │ vp9         279k video only              DASH video, webm_dash
3  webm 852x356     25 │  512k https │ vp9         512k video only              DASH video, webm_dash
4  webm 1280x534    25 │ 1800k https │ vp9        1800k video only              DASH video, webm_dash

--------------------------

C:\Users\Administrator>youtube-dl -Uv -N 8 "https://vd196.mycdn.me/?expires=1646557772885&srcIp=116.xxx.xxx.161&pr=10&srcAg=GECKO&ms=45.136.22.4&type=4&sig=hLJz8ZaZSIk&ct=6&urls=185.226.53.6&clientType=0&id=2254678985272" --user-agent "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0" -o "test.mp4"
[debug] Command-line config: ['-Uv', '-N', '8', 'https://vd196.mycdn.me/?expires=1646557772885&srcIp=116.xxx.xxx.161&pr=10&srcAg=GECKO&ms=45.136.22.4&type=4&sig=hLJz8ZaZSIk&ct=6&urls=185.226.53.6&clientType=0&id=2254678985272', '--user-agent', 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0', '-o', 'test.mp4']
[debug] Encodings: locale cp1252, fs utf-8, out utf-8, err utf-8, pref cp1252
[debug] yt-dlp version 2022.02.04 [c1653e9] (win_exe)
[debug] Python version 3.8.10 (CPython 64bit) - Windows-10-10.0.19042-SP0
[debug] exe versions: ffmpeg 5.0-full_build-www.gyan.dev (setts), ffprobe 4.4.1-full_build-www.gyan.dev
[debug] Optional libraries: Cryptodome, mutagen, sqlite, websockets
[debug] Proxy map: {}
Latest version: 2022.02.04, Current version: 2022.02.04
yt-dlp is up to date (2022.02.04)
[debug] [generic] Extracting URL: https://vd196.mycdn.me/?expires=1646557772885&srcIp=116.xxx.xxx.161&pr=10&srcAg=GECKO&ms=45.136.22.4&type=4&sig=hLJz8ZaZSIk&ct=6&urls=185.226.53.6&clientType=0&id=2254678985272
[generic] ?expires=1646557772885&srcIp=116.xxx.xxx.161&pr=10&srcAg=GECKO&ms=45.136.22.4&type=4&sig=hLJz8ZaZSIk&ct=6&urls=185.226.53: Requesting header
WARNING: [generic] Falling back on generic information extractor.
[generic] ?expires=1646557772885&srcIp=116.xxx.xxx.161&pr=10&srcAg=GECKO&ms=45.136.22.4&type=4&sig=hLJz8ZaZSIk&ct=6&urls=185.226.53: Downloading webpage
[generic] ?expires=1646557772885&srcIp=116.xxx.xxx.161&pr=10&srcAg=GECKO&ms=45.136.22.4&type=4&sig=hLJz8ZaZSIk&ct=6&urls=185.226.53: Extracting information
[debug] Identified a DASH manifest
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
WARNING: Extractor failed to obtain "title". Creating a generic title instead
[debug] Default format spec: bestvideo*+bestaudio/best
[info] ?expires=1646557772885&srcIp=116.xxx.xxx.161&pr=10&srcAg=GECKO&ms=45.136.22.4&type=4&sig=hLJz8ZaZSIk&ct=6&urls=185.226.53: Downloading 1 format(s): 4+7
[debug] Invoking downloader on "https://vd196.mycdn.me/?expires=1646557772885&srcIp=116.xxx.xxx.161&pr=10&srcAg=GECKO&ms=45.136.22.4&type=3&sig=5zzC4L6DCJE&ct=21&urls=185.226.53.6&clientType=0&id=2254678985272"
[download] Resuming download at byte 6958108
[download] Destination: test.mp4.f4.webm
[download]   0.9% of 1.06GiB at 247.80KiB/s ETA 01:13:59

```

### LINKED ISSUE #4462: [ok.ru] cannot get all quality of video
### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2022.07.18** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [ ] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

_No response_

### Provide a description that is worded well enough to be understood

hello,
i can see many quality video in browser but yt-dlp can get only hd or sd in format list.
https://ok.ru/video/1510273649159
https://ok.ru/video/1951798069873

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[root@CentOS-79-64-minimal ~]# youtube-dl -UvF "https://ok.ru/video/1951798069873"
[debug] Command-line config: ['-UvF', 'https://ok.ru/video/1951798069873']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out UTF-8, error UTF-8, screen UTF-8
[debug] yt-dlp version 2022.07.18 [135f05ef6]
[debug] Lazy loading extractors is disabled
[debug] Python 3.7.13 (CPython 64bit) - Linux-3.10.0-1160.59.1.el7.x86_64-x86_64-with-centos-7.9.2009-Core (glibc 2.17)
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 5.0.1 (fdk,setts), ffprobe 5.0.1, phantomjs 2.1.1
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2022.06.15, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.07.18, Current version: 2022.07.18
yt-dlp is up to date (2022.07.18)
[debug] [Odnoklassniki] Extracting URL: https://ok.ru/video/1951798069873
[Odnoklassniki] 1951798069873: Downloading desktop webpage
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[info] Available formats for 1951798069873:
ID EXT RESOLUTION │ PROTO │ VCODEC  ACODEC
───────────────────────────────────────────
hd mp4 unknown    │ https │ unknown unknown

[root@CentOS-79-64-minimal ~]# youtube-dl -UvF "https://ok.ru/video/1510273649159"
[debug] Command-line config: ['-UvF', 'https://ok.ru/video/1510273649159']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out UTF-8, error UTF-8, screen UTF-8
[debug] yt-dlp version 2022.07.18 [135f05ef6]
[debug] Lazy loading extractors is disabled
[debug] Python 3.7.13 (CPython 64bit) - Linux-3.10.0-1160.59.1.el7.x86_64-x86_64-with-centos-7.9.2009-Core (glibc 2.17)
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 5.0.1 (fdk,setts), ffprobe 5.0.1, phantomjs 2.1.1
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2022.06.15, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.07.18, Current version: 2022.07.18
yt-dlp is up to date (2022.07.18)
[debug] [Odnoklassniki] Extracting URL: https://ok.ru/video/1510273649159
[Odnoklassniki] 1510273649159: Downloading desktop webpage
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[info] Available formats for 1510273649159:
ID EXT RESOLUTION │ PROTO │ VCODEC  ACODEC
───────────────────────────────────────────
sd mp4 unknown    │ https │ unknown unknown
```

### LINKED ISSUE #7201: [ok.ru] Can't download non hls formats
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting that yt-dlp is broken on a **supported** site
- [X] I've verified that I'm running yt-dlp version **2023.03.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [ ] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

_No response_

### Provide a description that is worded well enough to be understood

Can't download non hls videos.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['https://ok.ru/video/1868262214220', '-f', 'full', '-vU', '--ignore-config']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.03.04 [392389b7d] (source)
[debug] Lazy loading extractors is disabled
[debug] Git HEAD: f41b949a
[debug] Python 3.8.10 (CPython x86_64 64bit) - Linux-5.15.0-71-generic-x86_64-with-glibc2.29 (OpenSSL 1.1.1f  31 Mar 2020, glibc 2.31)
[debug] exe versions: ffmpeg 4.2.7, ffprobe 4.2.7
[debug] Optional libraries: certifi-2019.11.28, secretstorage-2.3.1, sqlite3-2.6.0
[debug] Proxy map: {}
[debug] Loaded 1842 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Available version: stable@2023.03.04, Current version: stable@2023.03.04
yt-dlp is up to date (stable@2023.03.04)
[Odnoklassniki] Extracting URL: https://ok.ru/video/1868262214220
[Odnoklassniki] 1868262214220: Downloading desktop webpage
[Odnoklassniki] 1868262214220: Downloading m3u8 information
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[info] 1868262214220: Downloading 1 format(s): full
[debug] Invoking http downloader on "https://vd292.mycdn.me/?expires=1685752437000&srcIp=114.10.4.70&pr=10&srcAg=CHROME&ms=185.226.52.42&type=5&sig=7FnX_Adcp0c&ct=0&urls=45.136.22.19&clientType=0&id=602557450828"
ERROR: unable to download video data: HTTP Error 400: Bad Request
Traceback (most recent call last):
  File "/home/nir/Code-Fu/lain-lain/yt-dlp/yt_dlp/YoutubeDL.py", line 3280, in process_info
    success, real_download = self.dl(temp_filename, info_dict)
  File "/home/nir/Code-Fu/lain-lain/yt-dlp/yt_dlp/YoutubeDL.py", line 3003, in dl
    return fd.download(name, new_info, subtitle)
  File "/home/nir/Code-Fu/lain-lain/yt-dlp/yt_dlp/downloader/common.py", line 456, in download
    ret = self.real_download(filename, info_dict)
  File "/home/nir/Code-Fu/lain-lain/yt-dlp/yt_dlp/downloader/http.py", line 377, in real_download
    establish_connection()
  File "/home/nir/Code-Fu/lain-lain/yt-dlp/yt_dlp/downloader/http.py", line 129, in establish_connection
    ctx.data = self.ydl.urlopen(request)
  File "/home/nir/Code-Fu/lain-lain/yt-dlp/yt_dlp/YoutubeDL.py", line 3776, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/lib/python3.8/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/usr/lib/python3.8/urllib/request.py", line 640, in http_response
    response = self.parent.error(
  File "/usr/lib/python3.8/urllib/request.py", line 569, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 400: Bad Request
```

### PR DESCRIPTION
**Pull request by bashonly**
- Better HLS extraction (check both keys in metadata JSON)
 - Better DASH extraction (check for MPD URL in addition to embedded manifest data)
 - Extract webm DASH formats if Firefox UA header is passed
 - Clear CDN cookies which cause direct http downloads to fail with http error 400
 
Closes #2959, Closes #4462, Closes #7201

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details><details><summary>Copilot Summary</summary>  

### Summary
🌐🍪🧪

Improve `OdnoklassnikiIE` extractor by adding MPD support, fixing cookie issue, and updating tests.

> _`OdnoklassnikiIE`_
> _Fixes errors, adds MPD_
> _Winter of changes_

### Walkthrough
*  Add support for MPD formats and clear cookies for direct HTTP downloads in `OdnoklassnikiIE` extractor ([link](https://github.com/yt-dlp/yt-dlp/pull/7217/files?diff=unified&w=0#diff-63d16b041af43e1631e7294a988893d4480a5ebfd2b4fbc3060505a8cf3a6073R238-R244), [link](https://github.com/yt-dlp/yt-dlp/pull/7217/files?diff=unified&w=0#diff-63d16b041af43e1631e7294a988893d4480a5ebfd2b4fbc3060505a8cf3a6073L367-R399), [link](https://github.com/yt-dlp/yt-dlp/pull/7217/files?diff=unified&w=0#diff-63d16b041af43e1631e7294a988893d4480a5ebfd2b4fbc3060505a8cf3a6073R417), [link](https://github.com/yt-dlp/yt-dlp/pull/7217/files?diff=unified&w=0#diff-63d16b041af43e1631e7294a988893d4480a5ebfd2b4fbc3060505a8cf3a6073R451-R454), [link](https://github.com/yt-dlp/yt-dlp/pull/7217/files?diff=unified&w=0#diff-63d16b041af43e1631e7294a988893d4480a5ebfd2b4fbc3060505a8cf3a6073L433-R462))
*  Use `traverse_obj` function to safely access metadata keys and `url_or_none` function to filter out invalid URLs in `OdnoklassnikiIE` extractor ([link](https://github.com/yt-dlp/yt-dlp/pull/7217/files?diff=unified&w=0#diff-63d16b041af43e1631e7294a988893d4480a5ebfd2b4fbc3060505a8cf3a6073L367-R399), [link](https://github.com/yt-dlp/yt-dlp/pull/7217/files?diff=unified&w=0#diff-63d16b041af43e1631e7294a988893d4480a5ebfd2b4fbc3060505a8cf3a6073R21))
*  Import `urllib.parse` module and `HEADRequest` class to parse and request CDN URLs in `OdnoklassnikiIE` extractor ([link](https://github.com/yt-dlp/yt-dlp/pull/7217/files?diff=unified&w=0#diff-63d16b041af43e1631e7294a988893d4480a5ebfd2b4fbc3060505a8cf3a6073R1-R2), [link](https://github.com/yt-dlp/yt-dlp/pull/7217/files?diff=unified&w=0#diff-63d16b041af43e1631e7294a988893d4480a5ebfd2b4fbc3060505a8cf3a6073R12))
*  Update test cases for `OdnoklassnikiIE` extractor to use regular expressions, different formats and user agents, and skip problematic downloads ([link](https://github.com/yt-dlp/yt-dlp/pull/7217/files?diff=unified&w=0#diff-63d16b041af43e1631e7294a988893d4480a5ebfd2b4fbc3060505a8cf3a6073L44-R48), [link](https://github.com/yt-dlp/yt-dlp/pull/7217/files?diff=unified&w=0#diff-63d16b041af43e1631e7294a988893d4480a5ebfd2b4fbc3060505a8cf3a6073L68-R79), [link](https://github.com/yt-dlp/yt-dlp/pull/7217/files?diff=unified&w=0#diff-63d16b041af43e1631e7294a988893d4480a5ebfd2b4fbc3060505a8cf3a6073L84-R96), [link](https://github.com/yt-dlp/yt-dlp/pull/7217/files?diff=unified&w=0#diff-63d16b041af43e1631e7294a988893d4480a5ebfd2b4fbc3060505a8cf3a6073L101-R114), [link](https://github.com/yt-dlp/yt-dlp/pull/7217/files?diff=unified&w=0#diff-63d16b041af43e1631e7294a988893d4480a5ebfd2b4fbc3060505a8cf3a6073L119-R129), [link](https://github.com/yt-dlp/yt-dlp/pull/7217/files?diff=unified&w=0#diff-63d16b041af43e1631e7294a988893d4480a5ebfd2b4fbc3060505a8cf3a6073L156-R166))
*  Remove outdated TODO comment in `OdnoklassnikiIE` extractor ([link](https://github.com/yt-dlp/yt-dlp/pull/7217/files?diff=unified&w=0#diff-63d16b041af43e1631e7294a988893d4480a5ebfd2b4fbc3060505a8cf3a6073L148))

</details>

### CODE REVIEW

File: yt_dlp/extractor/odnoklassniki.py [hunk 1]
```diff
@@ -226,6 +235,13 @@ class OdnoklassnikiIE(InfoExtractor):
         'skip': 'Site no longer embeds',
     }]
 
+    # Direct http downloads will fail if CDN cookies are set from extraction
+    def _clear_cookies(self, cdn_url):
```
Comment (Grub4K) [inline yt_dlp/extractor/odnoklassniki.py:239]:
```suggestion
    def _clear_cookies(self, cdn_url):
        # Direct http downloads will fail if CDN cookies are set
        # so we need to reset them after each format extraction
```

**Review by Grub4K — APPROVED** [review]

File: yt_dlp/extractor/odnoklassniki.py [hunk 2]
```diff
@@ -226,6 +235,14 @@ class OdnoklassnikiIE(InfoExtractor):
...
     }]
 
+    def _clear_cookies(self, cdn_url):
+        # Direct http downloads will fail if CDN cookies are set
+        # so we need to reset them after each format extraction
+        if self._get_cookies('https://notarealsubdomain.mycdn.me/'):
+            self.cookiejar.clear(domain='.mycdn.me')
+        if self._get_cookies(cdn_url):
+            self.cookiejar.clear(domain=urllib.parse.urlparse(cdn_url).netloc)
+
```
Comment (pukkandan) [inline yt_dlp/extractor/odnoklassniki.py:245]: I don't think the if should be needed here. But I'll investigate it separately.

Comment (bashonly) [reply yt_dlp/extractor/odnoklassniki.py:245]: If there are no cookies scoped to the domain argument, `clear` will keyerror

Comment (pukkandan) [reply yt_dlp/extractor/odnoklassniki.py:245]: I know. Saying we should improve that.

**Pull request merged**